File: core/src/main/java/com/larryhsiao/nyx/core/tags/TagsByKeyword.java
package com.larryhsiao.nyx.core.tags;

import com.silverhetch.clotho.Source;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;

/**
 * Source to build tags by Jot id.
 */
public class TagsByKeyword implements Source<ResultSet> {
    private final Source<Connection> connSource;
    private final String keyword;

    public TagsByKeyword(Source<Connection> connSource, String keyword) {
        this.connSource = connSource;
        this.keyword = keyword;
    }

    @Override
    public ResultSet value() {
        try {
            PreparedStatement stmt = connSource.value().prepareStatement(
                //language=H2
                "SELECT  * FROM TAGS " +
                    "WHERE TITLE LIKE ?;"
            );
            stmt.setString(1, "%" + keyword +"%");
            return stmt.executeQuery();
        } catch (Exception e) {
            throw new IllegalArgumentException(e);
        }
    }
}

File: core/src/main/java/com/larryhsiao/nyx/core/tags/TagsByKeyword.kt
package com.larryhsiao.nyx.core.tags

import com.larryhsiao.clotho.Source
import java.sql.Connection
import java.sql.ResultSet

/**
 * Source to build tags by Jot id.
 */
class TagsByKeyword(
    private val connSource: Source<Connection>,
    private val keyword: String
) : Source<ResultSet> {
    override fun value(): ResultSet {
        return try {
            val stmt = connSource.value().prepareStatement( //language=H2
                "SELECT  * FROM TAGS " +
                    "WHERE UPPER(TITLE) LIKE UPPER(?) AND DELETE = 0;"
            )
            stmt.setString(1, "%$keyword%")
            stmt.executeQuery()
        } catch (e: Exception) {
            throw IllegalArgumentException(e)
        }
    }
}